internal/js_library/js_library.bzl

```diff
@@ -50,95 +56,147 @@ def write_amd_names_shim(actions, amd_names_shim, targets):
+            # workspace which indicates we should include the provider.
+            include_npm_package_info = True
 
     files_depset = depset(files)
+    js_files_depset = depset(js_files)
+    named_module_srcs_depset = depset(ctx.files.named_module_srcs)
+    typings_depset = depset(typings)
+    sources_depsets = [files_depset]
+    transitive_files_depsets = [files_depset]
+
+    for dep in ctx.attr.deps:
+        if NpmPackageInfo in dep:
+            sources_depsets.append(dep[NpmPackageInfo].sources)
+        elif DefaultInfo in dep:
+            transitive_files_depsets.append(dep[DefaultInfo].files)
+
+    transitive_sources = depset(transitive = sources_depsets)
 
     providers = [
         DefaultInfo(
-            files = files_depset,
-            runfiles = ctx.runfiles(files = ctx.files.srcs),
+            files = depset(transitive = transitive_files_depsets),
+            runfiles = ctx.runfiles(
+                files = files,
+                transitive_files = depset(transitive = transitive_files_depsets),
+            ),
         ),
         AmdNamesInfo(names = ctx.attr.amd_names),
-        js_module_info(depset(js_files)),
+        js_module_info(
+            sources = js_files_depset,
+            deps = ctx.attr.deps,
+        ),
+        js_named_module_info(
+            sources = named_module_srcs_depset,
+            deps = ctx.attr.deps,
+        ),
     ]
 
     if ctx.attr.package_name:
         path = "/".join([p for p in [ctx.bin_dir.path, ctx.label.workspace_root, ctx.label.package] if p])
         providers.append(LinkablePackageInfo(
             package_name = ctx.attr.package_name,
             path = path,
-            files = files_depset,
+            files = depset([
+                files_depset,
+                named_module_srcs_depset,
+            ]),
+        ))
+
+    if include_npm_package_info:
+        workspace_name = ctx.label.workspace_name if ctx.label.workspace_name else ctx.workspace_name
+        providers.append(NpmPackageInfo(
+            direct_sources = files_depset,
+            sources = transitive_sources,
+            workspace = workspace_name,
         ))
 
     # Don't provide DeclarationInfo if there are no typings to provide.
     # Improves error messaging downstream if DeclarationInfo is required.
     if len(typings):
-        providers.append(declaration_info(depset(typings)))
+        providers.append(declaration_info(
+            declarations = typings_depset,
+            deps = ctx.attr.deps,
+        ))
 
     return providers
 
 _js_library = rule(
     implementation = _impl,
     attrs = {
-        "amd_names": attr.string_dict(doc = _AMD_NAMES_DOC),
-        "is_windows": attr.bool(mandatory = True, doc = "Automatically set by macro"),
+        "amd_names": attr.string_dict(
+            doc = _AMD_NAMES_DOC,
+        ),
+        "deps": attr.label_list(
+            doc = "Transitive dependencies of the package",
```

Comment:
we should explain a bit more here based on discussion today

Reply:
I've included a more complete description for deps